### PR TITLE
secure_storage: preferred way of programming RPMB key

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Describe the preferred way of programming the RPMB key and add risks of automatic programming from OP-TEE.

### CODE REVIEW

File: architecture/secure_storage.rst [hunk 1]
```diff
@@ -380,6 +385,29 @@ platforms shall use to implement its security logic
    For OP-TEE to be able to write the RPMB key, ``CFG_RPMB_WRITE_KEY=y`` must be
    configured and ``plat_rpmb_key_is_ready()`` must allow it at runtime.
 
+When programming the RPMB key from normal world the RPMB key must be made
+available to that tool. The RPMB key must for security reasons normally not
+be known outside OP-TEE, but an exception might be made during production.
+
+If the HUK is known the script `scripts/derive_rpmb_key.py`_ can be used to
```
Comment (jbech-linaro) [inline architecture/secure_storage.rst:392]:
```suggestion
If the HUK is known, the script `scripts/derive_rpmb_key.py`_ can be used to
```

**Review by jbech-linaro — COMMENTED** [review]
Minor fix + open questions, other than that:
`Reviewed-by: Joakim Bech <joakim.bech@linaro.org>`

File: architecture/secure_storage.rst [hunk 2]
```diff
@@ -380,6 +385,29 @@ platforms shall use to implement its security logic
...
+derive the RPMB key.
+
+Pros and cons with OP-TEE automatically writing the RPMB key:
+    1. Automatic writing can be triggered after each boot controlled
+       entirely by the normal world, essentially tricking the secure world
+       to reveal the secret RPMB key. Having a separate OP-TEE binary for
+       production is not fully secure since it is enough to restore it from
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply architecture/secure_storage.rst:399]: "It" is the binary used at the factory. If you store it in non-volatile memory, it's basically game over, but even if you can avoid that, too many people have access to a class break attack.

By production I meant what you do in the factory, I'll rephrase it to make it clear.

File: architecture/secure_storage.rst [hunk 3]
```diff
@@ -380,6 +385,29 @@ platforms shall use to implement its security logic
...
+       from one device to achieve class breakage.
+
+    2. Automatic writing may target the wrong device if there is more than
+       one RPMB since the device name it determined by the probe order. The
+       probe order may differ between boot loader and the kernel or when
+       rebooting.
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply architecture/secure_storage.rst:405]: OK, I'll update.

File: architecture/secure_storage.rst [hunk 4]
```diff
@@ -380,6 +385,29 @@ platforms shall use to implement its security logic
...
 
+When programming the RPMB key from normal world the RPMB key must be made
+available to that tool. The RPMB key must for security reasons normally not
+be known outside OP-TEE, but an exception might be made during production.
+
+If the HUK is known the script `scripts/derive_rpmb_key.py`_ can be used to
+derive the RPMB key.
+
+Pros and cons with OP-TEE automatically writing the RPMB key:
+    1. Automatic writing can be triggered after each boot controlled
```
[the remark this answers is not in the document]
Comment (jbech-linaro) [reply architecture/secure_storage.rst:396]: Yes, I think that makes sense, I.e, anything that makes is clear to others that using this in production is a no-no.

**Comment by jenswi-linaro** [comment]
Update

File: architecture/secure_storage.rst [hunk 5]
```diff
@@ -387,22 +390,24 @@ platforms shall use to implement its security logic
...
-       production is not fully secure since it is enough to restore it from
+       to reveal the secret RPMB key. Having a separate OP-TEE binary in the
+       factory is not fully secure since it is enough to restore it from
        from one device to achieve class breakage.
 
     2. Automatic writing may target the wrong device if there is more than
        one RPMB since the device name it determined by the probe order. The
        probe order may differ between boot loader and the kernel or when
-       rebooting.
+       rebooting. This problem can be addressed by supplying -rpmb-cid and
```
Comment (jforissier) [inline architecture/secure_storage.rst:409]: That's `--rpmb-cid` with two dashes.

**Review by jforissier — COMMENTED** [review]
With my comment below addressed:

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: architecture/secure_storage.rst (hunk 5 above)
Comment (jenswi-linaro) [reply architecture/secure_storage.rst:409]: Yes, thanks.

**Comment by jenswi-linaro** [comment]
Comment address, squashed and tags applied.

File: architecture/secure_storage.rst [hunk 6]
```diff
@@ -341,14 +341,22 @@ on the device: this is a one time action for the lifetime of the device. Once
 the key has been written on the eMMC controller, the controller uses it to
 authenticate requests.
 
-This key can be programmed in a number of ways. If you want OP-TEE to do it,
+The RPMB key can be programmed in a number of ways. The safest and most
+secure way is to program it from normal world during production in a secure
+environment. MMC-tools_ can for instance be used to program the RPMB
```
Comment (etienne-lms) [inline architecture/secure_storage.rst:346]: Misses inverse quotes around `MMC-tools_`?

File: architecture/secure_storage.rst [hunk 7]
```diff
@@ -380,6 +388,31 @@ platforms shall use to implement its security logic
...
+    2. Automatic writing may target the wrong device if there is more than
+       one RPMB since the device name it determined by the probe order. The
+       probe order may differ between boot loader and the kernel or when
+       rebooting. This problem can be addressed by supplying --rpmb-cid and
+       the CID of the MMC device to use as argument.
+
+    3. Automatic writing must be used at the moment if tee-supplicant is
+       configured to emulate RPMB since it starts from scratch at each
+       boot.
+
```
Comment (etienne-lms) [inline architecture/secure_storage.rst:415]: can remove the extra empty line

File: architecture/secure_storage.rst [hunk 8]
```diff
@@ -380,6 +388,31 @@ platforms shall use to implement its security logic
...
+       factory is not fully secure since it is enough to restore it from
+       from one device to achieve class breakage.
+
+    2. Automatic writing may target the wrong device if there is more than
+       one RPMB since the device name it determined by the probe order. The
+       probe order may differ between boot loader and the kernel or when
+       rebooting. This problem can be addressed by supplying --rpmb-cid and
+       the CID of the MMC device to use as argument.
+
+    3. Automatic writing must be used at the moment if tee-supplicant is
```
Comment (etienne-lms) [inline architecture/secure_storage.rst:412]: remove `if`?
"... must be used at the moment tee-supplicant is configured to emulate..."

File: architecture/secure_storage.rst [hunk 9]
```diff
@@ -461,3 +494,5 @@ specification.
 .. _core/tee/tee_fs_key_manager.c: https://github.com/OP-TEE/optee_os/blob/master/core/tee/tee_fs_key_manager.c
 .. _core/tee/tee_rpmb_fs.c: https://github.com/OP-TEE/optee_os/blob/master/core/tee/tee_rpmb_fs.c
 .. _lib/libutee/: https://github.com/OP-TEE/optee_os/blob/master/lib/libutee/
+.. _mmc-tools: https://docs.kernel.org/driver-api/mmc/mmc-tools.html
```
Comment (etienne-lms) [inline architecture/secure_storage.rst:497]: Refs are not case sensitive?  (`MMC-tools_` at line 346  vs `_mmc-tools` here)

File: architecture/secure_storage.rst (hunk 8 above)
Comment (jenswi-linaro) [reply architecture/secure_storage.rst:412]: I'm sorry, I don't follow.

File: architecture/secure_storage.rst (hunk 7 above)
Comment (jenswi-linaro) [reply architecture/secure_storage.rst:415]: OK

File: architecture/secure_storage.rst (hunk 6 above)
Comment (jenswi-linaro) [reply architecture/secure_storage.rst:346]: Why should we have that? We don't use it with mmc-utils_.

File: architecture/secure_storage.rst (hunk 9 above)
Comment (jenswi-linaro) [reply architecture/secure_storage.rst:497]: It seems it's not. I can make it _MMC-tools instead if you prefer that.

**Comment by jenswi-linaro** [comment]
Squashed and ready.
